File: python/src/venn_diagram_lab/statistics.py
"""Statistical metrics for Venn diagram region analysis.

Ported from the web tool (src/utils/statistics.ts) to maintain numerical parity.
All functions are pure and operate on primitive scalar inputs (set sizes,
intersection counts) so they can be unit-tested in isolation.
"""

from __future__ import annotations

from collections.abc import Mapping
from dataclasses import dataclass
from itertools import combinations

import numpy as np
import pandas as pd
from scipy.stats import false_discovery_control, hypergeom


def jaccard(size_a: int, size_b: int, intersection: int) -> float:
    """Jaccard similarity: |A intersect B| / |A union B|.

    Returns 0.0 if both sets are empty (convention matches the web tool).
    """
    union = size_a + size_b - intersection
    if union <= 0:
        return 0.0
    return intersection / union


def dice(size_a: int, size_b: int, intersection: int) -> float:
    """Dice (Sørensen-Dice) coefficient: 2|A ∩ B| / (|A| + |B|)."""
    denom = size_a + size_b
    if denom <= 0:
        return 0.0
    return (2.0 * intersection) / denom


def overlap_coefficient(size_a: int, size_b: int, intersection: int) -> float:
    """Szymkiewicz-Simpson overlap coefficient: |A intersect B| / min(|A|, |B|)."""
    denom = min(size_a, size_b)
    if denom <= 0:
        return 0.0
    return intersection / denom


def hypergeometric_p_value(N: int, K: int, n: int, k: int) -> float:  # noqa: N803
    """One-sided over-representation p-value: P(X >= k) where X ~ Hypergeometric(N, K, n).

    Parameters
    ----------
    N : population size (total items in the universe)
    K : number of "success" states in the population (e.g. |A| inclusive)
    n : number of draws (e.g. |B| inclusive)
    k : observed successes (e.g. |A intersect B| inclusive)

    Returns 1.0 for invalid inputs (matches web tool convention) so the metric
    is safe to feed into BH-FDR without filtering.
    """
    if N < 1 or K < 0 or n < 0 or k < 0:
        return 1.0
    K = min(K, N)  # noqa: N806
    n = min(n, N)
    if k > min(K, n):
        return 1.0
    # P(X >= k) = P(X > k - 1) = sf(k - 1, N, K, n)
    p = float(hypergeom.sf(k - 1, N, K, n))
    return min(max(p, 0.0), 1.0)


def fold_enrichment(N: int, K: int, n: int, k: int) -> float:  # noqa: N803
    """Fold enrichment: observed / expected = (k * N) / (K * n).

    Returns 0.0 if any denominator is zero (matches web tool convention).
    """
    if N == 0 or K == 0 or n == 0:
        return 0.0
    return (k * N) / (K * n)


def bh_fdr(p_values: list[float]) -> list[float]:
    """Benjamini-Hochberg FDR adjustment, preserving input order.

    Returns adjusted p-values in the same order as the input. Empty input -> empty output.
    """
    if not p_values:
        return []
    adjusted = false_discovery_control(p_values, method="bh")
    # Clip into [0, 1] for safety (scipy already does this, but be explicit).
    return [min(max(float(p), 0.0), 1.0) for p in adjusted]


# Significance thresholds (match web tool defaults)
_THRESHOLD_SIGNIFICANT = 0.05
_THRESHOLD_HIGHLY_SIGNIFICANT = 0.001


@dataclass(frozen=True)
class StatisticsResult:
    """Container for the five pairwise metric tables produced by :func:`compute_pairwise`.

    All four similarity/enrichment matrices are square (NxN) DataFrames indexed
    and columned by set names — easy to feed into a heatmap renderer. The
    hypergeometric table is long-form (one row per pair) because each row carries
    multiple columns (intersection, expected, p_value, p_adjusted, flags).

    Attributes:
        jaccard: NxN Jaccard similarity matrix.
        dice: NxN Sørensen-Dice coefficient matrix.
        overlap_coefficient: NxN overlap coefficient (Szymkiewicz-Simpson).
        fold_enrichment: NxN fold-enrichment matrix.
        hypergeometric: Long-form table with set_a, set_b, intersection, expected,
            p_value, p_adjusted (Benjamini-Hochberg), significant, highly_significant.
    """

    jaccard: pd.DataFrame
    dice: pd.DataFrame
    overlap_coefficient: pd.DataFrame
    fold_enrichment: pd.DataFrame
    hypergeometric: pd.DataFrame


def _square_metric(
    set_names: list[str],
    pair_to_value: Mapping[tuple[str, str], float],
    diagonal_value: float,
) -> pd.DataFrame:
    n = len(set_names)
    matrix = np.full((n, n), diagonal_value, dtype=float)
    name_to_idx = {name: i for i, name in enumerate(set_names)}
    for (a, b), v in pair_to_value.items():
        i, j = name_to_idx[a], name_to_idx[b]
        matrix[i, j] = v
        matrix[j, i] = v
    return pd.DataFrame(matrix, index=set_names, columns=set_names)


def compute_pairwise(
    set_names: list[str],
    inclusive_sizes: Mapping[str, int],
    pairwise_intersections: Mapping[tuple[str, str], int],
    universe_size: int,
) -> StatisticsResult:
    """Compute all five pairwise metric tables.

    Parameters
    ----------
    set_names : ordered list of set identifiers (e.g. ["SetA", "SetB"]).
    inclusive_sizes : map set name -> |set| (the inclusive count).
    pairwise_intersections : map (name_a, name_b) with name_a appearing earlier
        in set_names than name_b -> |set_a & set_b|. The order matters only for
        the long-form hypergeometric table (where set_a is always the
        earlier-listed set).
    universe_size : N for the hypergeometric test (total unique items across
        the dataset, NOT the number of rows when in aggregated mode).
    """
    pair_jaccard: dict[tuple[str, str], float] = {}
    pair_dice: dict[tuple[str, str], float] = {}
    pair_oc: dict[tuple[str, str], float] = {}
    pair_fe: dict[tuple[str, str], float] = {}

    rows: list[dict[str, object]] = []
    raw_p_values: list[float] = []

    for a, b in combinations(set_names, 2):
        ka = inclusive_sizes[a]
        kb = inclusive_sizes[b]
        inter = pairwise_intersections.get((a, b), pairwise_intersections.get((b, a), 0))

        pair_jaccard[(a, b)] = jaccard(ka, kb, inter)
        pair_dice[(a, b)] = dice(ka, kb, inter)
        pair_oc[(a, b)] = overlap_coefficient(ka, kb, inter)
        pair_fe[(a, b)] = fold_enrichment(universe_size, ka, kb, inter)

        expected = (ka * kb) / universe_size if universe_size > 0 else 0.0
        p_val = hypergeometric_p_value(universe_size, ka, kb, inter)
        raw_p_values.append(p_val)

        rows.append({
            "set_a": a,
            "set_b": b,
            "intersection": inter,
            "expected": expected,
            "p_value": p_val,
        })

    adjusted = bh_fdr(raw_p_values)
    for row, adj in zip(rows, adjusted, strict=True):
        row["p_adjusted"] = adj
        row["significant"] = bool(adj < _THRESHOLD_SIGNIFICANT)
        row["highly_significant"] = bool(adj < _THRESHOLD_HIGHLY_SIGNIFICANT)

    hyp_dtypes = {
        "set_a": "string",
        "set_b": "string",
        "intersection": "int64",
        "expected": "float64",
        "p_value": "float64",
        "p_adjusted": "float64",
        "significant": "bool",
        "highly_significant": "bool",
    }
    if rows:
        hyp_df = pd.DataFrame(rows).astype(hyp_dtypes)
    else:
        hyp_df = pd.DataFrame({col: pd.Series(dtype=dt) for col, dt in hyp_dtypes.items()})

    hyp_df = hyp_df.sort_values("p_value", kind="mergesort", ignore_index=True)

    return StatisticsResult(
        jaccard=_square_metric(set_names, pair_jaccard, diagonal_value=1.0),
        dice=_square_metric(set_names, pair_dice, diagonal_value=1.0),
        overlap_coefficient=_square_metric(set_names, pair_oc, diagonal_value=1.0),
        fold_enrichment=_square_metric(set_names, pair_fe, diagonal_value=float("nan")),
        hypergeometric=hyp_df,
    )

File: r/R/statistics.R
#' Jaccard similarity index
#'
#' Computes |A intersection B| / |A union B|. Matches the web tool's convention
#' of returning 0 when both sets are empty (NaN-safe).
#'
#' @param size_a Inclusive size of set A (integer >= 0).
#' @param size_b Inclusive size of set B (integer >= 0).
#' @param intersection Inclusive intersection size |A intersection B|.
#' @return Numeric in [0, 1].
#' @export
#' @examples
#' jaccard(10, 10, 5)
#' jaccard(0, 0, 0)
jaccard <- function(size_a, size_b, intersection) {
    union_size <- size_a + size_b - intersection
    if (union_size <= 0) return(0)
    intersection / union_size
}

#' Sorensen-Dice coefficient
#'
#' Computes 2 * |A intersection B| / (|A| + |B|). Returns 0 if both sets are
#' empty (matches web tool convention).
#'
#' @inheritParams jaccard
#' @return Numeric in [0, 1].
#' @export
#' @examples
#' dice(10, 10, 5)
dice <- function(size_a, size_b, intersection) {
    denom <- size_a + size_b
    if (denom <= 0) return(0)
    (2 * intersection) / denom
}

#' Szymkiewicz-Simpson overlap coefficient
#'
#' Computes |A intersection B| / min(|A|, |B|). Useful when one set is much
#' smaller than the other.
#'
#' @inheritParams jaccard
#' @return Numeric in [0, 1].
#' @export
#' @examples
#' overlap_coefficient(10, 5, 3)
overlap_coefficient <- function(size_a, size_b, intersection) {
    denom <- min(size_a, size_b)
    if (denom <= 0) return(0)
    intersection / denom
}

#' One-sided hypergeometric p-value (over-representation)
#'
#' Computes P(X >= k) where X ~ Hypergeometric(N, K, n). Returns 1.0 for
#' invalid inputs so the metric is safe to feed into BH-FDR without filtering.
#'
#' Maps to R's `phyper(k - 1, K, N - K, n, lower.tail = FALSE)`. Note that R's
#' phyper parameter convention differs from Python's scipy: R uses `m` for
#' success-in-population and `n` for failure-in-population (= N - K), where
#' Python uses `N` for total population.
#'
#' @param N Population size (total items in the universe). Integer >= 1.
#' @param K Number of success states in the population (e.g. inclusive |A|). Integer >= 0.
#' @param n Number of draws (e.g. inclusive |B|). Integer >= 0.
#' @param k Observed successes (e.g. |A intersection B|). Integer >= 0.
#' @return Numeric in [0, 1].
#' @export
#' @examples
#' hypergeometric_p_value(20000, 138, 581, 126)
hypergeometric_p_value <- function(N, K, n, k) {
    if (N < 1 || K < 0 || n < 0 || k < 0) return(1.0)
    K_clipped <- min(K, N)
    n_clipped <- min(n, N)
    if (k > min(K_clipped, n_clipped)) return(1.0)
    p <- stats::phyper(k - 1, K_clipped, N - K_clipped, n_clipped, lower.tail = FALSE)
    min(max(p, 0.0), 1.0)
}

#' Fold enrichment (observed / expected ratio)
#'
#' Computes (k * N) / (K * n). Returns 0.0 if any denominator is zero
#' (matches web tool convention).
#'
#' @inheritParams hypergeometric_p_value
#' @return Numeric (>= 0; can exceed 1 for over-representation).
#' @export
#' @examples
#' fold_enrichment(20000, 138, 581, 126)
fold_enrichment <- function(N, K, n, k) {
    if (N == 0 || K == 0 || n == 0) return(0)
    (k * N) / (K * n)
}

#' Benjamini-Hochberg FDR adjustment
#'
#' Wraps `stats::p.adjust(p, method = "BH")`. Returns adjusted p-values in the
#' same order as the input. Empty input -> empty output.
#'
#' @param p_values Numeric vector of raw p-values in [0, 1].
#' @return Numeric vector of adjusted p-values, same length as input.
#' @export
#' @examples
#' bh_fdr(c(0.001, 0.01, 0.05, 0.5))
bh_fdr <- function(p_values) {
    if (length(p_values) == 0) return(numeric(0))
    stats::p.adjust(p_values, method = "BH")
}

# Internal: build a square N x N named matrix from (set_a, set_b) -> value entries.
.square_metric <- function(set_names, pair_values, diagonal_value) {
    n <- length(set_names)
    m <- matrix(diagonal_value, nrow = n, ncol = n,
                dimnames = list(set_names, set_names))
    for (key in names(pair_values)) {
        parts <- strsplit(key, "|", fixed = TRUE)[[1]]
        a <- parts[1L]; b <- parts[2L]
        m[a, b] <- pair_values[[key]]
        m[b, a] <- pair_values[[key]]
    }
    m
}

#' Compute all 5 pairwise statistical tables
#'
#' Orchestrator that returns a [`StatisticsResult-class`] populated with
#' Jaccard, Dice, Overlap Coefficient, Fold Enrichment (square NxN matrices)
#' plus a long-form hypergeometric table with BH-FDR adjustment.
#'
#' @param set_names Ordered character vector of set identifiers (e.g. c("A","B","C")).
#' @param inclusive_sizes Named integer vector of inclusive set sizes (`names(inclusive_sizes)` matches `set_names`).
#' @param pairwise_intersections Named list of pair intersection counts. Keys are
#'   "set_a|set_b" with set_a appearing earlier in `set_names` than set_b.
#' @param universe_size Hypergeometric universe N (population size). Integer >= 1.
#' @return A [`StatisticsResult-class`] object.
#' @export
#' @examples
#' compute_pairwise(
#'     set_names = c("A", "B"),
#'     inclusive_sizes = c(A = 10L, B = 8L),
#'     pairwise_intersections = list("A|B" = 5L),
#'     universe_size = 100L
#' )
compute_pairwise <- function(set_names, inclusive_sizes, pairwise_intersections, universe_size) {
    pair_jaccard <- list()
    pair_dice    <- list()
    pair_oc      <- list()
    pair_fe      <- list()

    rows_set_a        <- character()
    rows_set_b        <- character()
    rows_intersection <- integer()
    rows_expected     <- numeric()
    rows_p_value      <- numeric()

    pairs <- utils::combn(set_names, 2, simplify = FALSE)
    for (pair in pairs) {
        a <- pair[1L]; b <- pair[2L]
        ka <- inclusive_sizes[[a]]
        kb <- inclusive_sizes[[b]]
        # Look up the intersection count under "a|b" or fall back to "b|a".
        key_ab <- paste(a, b, sep = "|")
        key_ba <- paste(b, a, sep = "|")
        inter <- pairwise_intersections[[key_ab]]
        if (is.null(inter)) inter <- pairwise_intersections[[key_ba]]
        if (is.null(inter)) inter <- 0L
        inter <- as.integer(inter)

        pair_jaccard[[key_ab]] <- jaccard(ka, kb, inter)
        pair_dice[[key_ab]]    <- dice(ka, kb, inter)
        pair_oc[[key_ab]]      <- overlap_coefficient(ka, kb, inter)
        pair_fe[[key_ab]]      <- fold_enrichment(universe_size, ka, kb, inter)

        expected <- if (universe_size > 0) (ka * kb) / universe_size else 0
        p_val <- hypergeometric_p_value(universe_size, ka, kb, inter)

        rows_set_a        <- c(rows_set_a, a)
        rows_set_b        <- c(rows_set_b, b)
        rows_intersection <- c(rows_intersection, inter)
        rows_expected     <- c(rows_expected, expected)
        rows_p_value      <- c(rows_p_value, p_val)
    }

    adjusted <- bh_fdr(rows_p_value)
    significant        <- adjusted < 0.05
    highly_significant <- adjusted < 0.001

    hyp <- data.frame(
        set_a              = rows_set_a,
        set_b              = rows_set_b,
        intersection       = rows_intersection,
        expected           = rows_expected,
        p_value            = rows_p_value,
        p_adjusted         = adjusted,
        significant        = significant,
        highly_significant = highly_significant,
        stringsAsFactors   = FALSE
    )
    # Sort by p_value ascending (shell sort — stable for named vectors, deterministic).
    hyp <- hyp[order(hyp$p_value, method = "shell"), , drop = FALSE]
    rownames(hyp) <- NULL

    methods::new("StatisticsResult",
        jaccard             = .square_metric(set_names, pair_jaccard, diagonal_value = 1.0),
        dice                = .square_metric(set_names, pair_dice,    diagonal_value = 1.0),
        overlap_coefficient = .square_metric(set_names, pair_oc,      diagonal_value = 1.0),
        fold_enrichment     = .square_metric(set_names, pair_fe,      diagonal_value = NA_real_),
        hypergeometric      = hyp
    )
}
